### ISSUE TITLE
How to cancel blockwise PUT mid-flight

### ISSUE THREAD

**Issue by daFritz84** [eclipse/californium#2046]
Hey there,

I'm fiddling around with blockwise transfers, to be precise, with a Block1 transfer putting files on an embedded coap server (libaro). Sometimes I would like to cancel ongoing (long duration) file transfers mid-flight. Is there any method/mechanism which would achieve the desired behavior?

Btw. I'm using californium 3.6.0 on Android. The `cancel` method from `Request` class seems to work only for Observers.

Thanks in advance for any help,
Stefan

**Comment by boaks** [comment]
I would have assumed, that `Request.cancel()` would do the job, at least after the transfer of the next block.

I'm not sure, if I can have a look at this this week, but next week I will have more time.

**Comment by daFritz84** [comment]
Thanks @boaks,

I poked around in `BlockwiseLayer.java` when I encountered the following TODO comment:
https://github.com/eclipse/californium/blob/3eecb4c739a40ac1af066a89d863be338fa54e00/californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java#L169

From my perspective, I would expect that the `cancel` flag is evaluated upon sending the next block, like in https://github.com/eclipse/californium/blob/3eecb4c739a40ac1af066a89d863be338fa54e00/californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java#L1058

But unfortunately, there is only a test for the `completed` flag. At least that's as far as I got myself.

**Comment by boaks** [comment]
I tried to reproduce it, but didn't succeed. See commit #d62f1f9ecef8d0021187f55c4b1b2b40e60a2b8d

If `Request.cancel()` is called, that ends up in the `CleanupMessageObserver`, which completes the exchange.
Additionally, `BlockwiseLayer.handleBlock1Response` check also for `Request.isCanceled()`.

With that, I'm not sure, what's ongoing in your case.

**Comment by daFritz84** [comment]
Thank you for looking into the issue.

Transfers are working fine otherwise, so it might be some cross-thread-call issue. Or another case might be a retransmission loop (known bug that can occur in our implementation), when the coap server is unable to store the file and therefore constantly requests block 0 again and again instead of failing.

Anyways, I'll take a closer look at the problem next week and report back.

The afore mentioned TODO comment in BlockwiseLayer.java has left me in suspicion that the feature is missing. Can you shed any light what timeouts are not forwared or where cancellation is still missing?

**Comment by boaks** [comment]
"BLOCK1" and "requests block 0" ... not sure what your doing.

 > coap server (libaro).

So the coap-server is libaro and the coap-client is Californium?

So the server sends responses (e.g. block1 continue), but it can't request a block1.

> The afore mentioned TODO comment in BlockwiseLayer.java has left me in suspicion that the feature is missing. 
> Can you shed any light what timeouts are not forwared or where cancellation is still missing?

Many of that TODOs are somehow stale. And I'm mostly not the author, so it's hard to find out, what they mean.
 
I will remove the mentioned cancel, but I have no clue, what the mentioned timeout should mean. 
The coap-timeouts are applied for each single block-exchange. And the coap-client's response timeout is an additional timeout for the complete transfer only used by that coap-client. (For blockwise, I have my doubts about the usability of that.)

**Comment by boaks** [comment]
(I removed the stale TODOs in BlockwiseLayer.)

**Comment by boaks** [comment]
Maybe you can add some logging on your side to see, if this is caused by an retry there. If so , you may check in your retry, it the `Request.isCanceled()`.

**Comment by daFritz84** [comment]
**TL;DR:** Thanks for the clarification, the stale comment got me on the wrong track. The bug was specific to our implementation, i.e cross thread call (ui thread -> coap worker) resulting in a lock that prevented the call to cancel until after the  transfer was finished.

So to clear up some confusion: 
One of the scenarios where I suspected potential problems was our current state of implementation. We do a blockwise transfer, PUT with block1 option. Until recently, when we were unable to store the data received by afore mentioned PUT, the store procedure failed which resulted in a retransmit request. At worst this could result in a retransmit loop of the first block (referred to Block 0 in my last comment).  Of course this is an obvious bug, and we should fail with the appropriate error code. 
I hope this makes a little more sense.

**Comment by boaks** [comment]
> Until recently, when we were unable to store the data received by afore mentioned PUT, the store procedure failed which resulted in a retransmit request.

Just to mention: if this delay is just a small couple of seconds, you may try, if [RFC7252 - Separate Response](https://datatracker.ietf.org/doc/html/rfc7252#section-5.2.2) helps you out. With that, send the ACK when you detect, that the write may take longer, and the response on success.
